### PR TITLE
Create API for submitting attendance

### PR DESCRIPTION
**Pull request by ghulamhafeez**
### Description

Create API for submitting member attendance
### Implementation

* Create server action to post the member attendance
* Unit Testing

### Screenshots

- Unit Test

Test

- Supabase Attendance Table 

Table

### CODE REVIEW

**Review by muhammad-bilal-ali-khokhar — APPROVED** [review]

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (Mubashar1009) [inline src/app/attendance/components/AttendanceFormBuilder.tsx:192]: can we use action instead of onSubmit

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (Mubashar1009) [inline src/app/attendance/components/AttendanceFormBuilder.tsx:251]: can we use hasError prop of input for handling error ?

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (Mubashar1009) [inline src/app/attendance/components/AttendanceFormBuilder.tsx:274]: here also

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (nadirhussain786) [inline src/app/attendance/components/AttendanceFormBuilder.tsx:184]: @ghulamhafeez  should return error message

File: src/app/attendance/actions/attendanceActions.test.tsx
Comment (Monfernape) [inline src/app/attendance/actions/attendanceActions.test.tsx:1]: File name is not correct. we're over contextualizing. Since this is already under actions directory, it can simply be called `createAttendance`

File: src/app/attendance/actions/attendanceActions.test.tsx
Comment (Monfernape) [inline src/app/attendance/actions/attendanceActions.test.tsx:4]: Can we write integration test? A test where we render the `AttendanceFormBuilder` and we render inputs and verify this function is called correctly?

File: src/app/attendance/actions/attendanceActions.test.tsx
Comment (Monfernape) [inline src/app/attendance/actions/attendanceActions.test.tsx:14]: typo.

File: src/app/attendance/actions/attendanceActions.tsx
Comment (Monfernape) [inline src/app/attendance/actions/attendanceActions.tsx:1]: File name please

File: src/app/attendance/actions/attendanceActions.tsx
Comment (Monfernape) [inline src/app/attendance/actions/attendanceActions.tsx:5]: This should be coming from routes. We've enum defined there.

File: src/app/attendance/actions/attendanceActions.tsx
Comment (Monfernape) [inline src/app/attendance/actions/attendanceActions.tsx:3]: This is relatlive import. Let's use absolute paths. See the above one.

File: src/app/attendance/actions/attendanceActions.tsx
Comment (Monfernape) [inline src/app/attendance/actions/attendanceActions.tsx:7]: Where's this coming from? Shouldn't it be coming from Attendance component?

File: src/app/attendance/actions/attendanceActions.tsx
Comment (Monfernape) [inline src/app/attendance/actions/attendanceActions.tsx:18]: Nice

File: src/app/attendance/actions/attendanceActions.tsx
Comment (Monfernape) [reply src/app/attendance/actions/attendanceActions.tsx:25]: @Mubashar1009 is correct here though. Let's use enums.

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (Monfernape) [inline src/app/attendance/components/AttendanceFormBuilder.tsx:172]: Why do we need to reset the form? If we just navigate away, it should be enough.

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (Monfernape) [inline src/app/attendance/components/AttendanceFormBuilder.tsx:181]:
```suggestion
        title: "Attendance could not be marked",
```

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (Monfernape) [inline src/app/attendance/components/AttendanceFormBuilder.tsx:190]: Why form within form?

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (Monfernape) [inline src/app/attendance/components/AttendanceFormBuilder.tsx:221]: This seems wrong. We should use placeholder instead.

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (Monfernape) [inline src/app/attendance/components/AttendanceFormBuilder.tsx:206]: This doesn't look correct. Why need to transform into string? Isn't it already string?

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (Monfernape) [inline src/app/attendance/components/AttendanceFormBuilder.tsx:208]: We should use `UserRole` enum.

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (Monfernape) [inline src/app/attendance/components/AttendanceFormBuilder.tsx:226]: Why `toString`?

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (Monfernape) [inline src/app/attendance/components/AttendanceFormBuilder.tsx:248]: What's going on here?

File: src/constant/constant.ts
Comment (Monfernape) [inline src/constant/constant.ts:21]: We don't need that. Route should be defined in routes enum.

File: src/types/index.tsx
Comment (Monfernape) [inline src/types/index.tsx:51]: @muhammad-bilal-ali-khokhar see if this is duplicate.

**Review by Monfernape — CHANGES_REQUESTED** [review]

File: src/app/attendance/actions/attendanceActions.tsx
Comment (ghulamhafeez) [reply src/app/attendance/actions/attendanceActions.tsx:25]: Fix

File: src/app/attendance/actions/attendanceActions.test.tsx
Comment (ghulamhafeez) [reply src/app/attendance/actions/attendanceActions.test.tsx:1]: Fix

File: src/app/attendance/actions/attendanceActions.tsx
Comment (ghulamhafeez) [reply src/app/attendance/actions/attendanceActions.tsx:1]: Fix

File: src/app/attendance/actions/attendanceActions.tsx
Comment (ghulamhafeez) [reply src/app/attendance/actions/attendanceActions.tsx:5]: Fix

File: src/app/attendance/actions/attendanceActions.tsx
Comment (ghulamhafeez) [reply src/app/attendance/actions/attendanceActions.tsx:3]: FIx

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (ghulamhafeez) [reply src/app/attendance/components/AttendanceFormBuilder.tsx:172]: Okay, Fix

File: src/constant/constant.ts
Comment (ghulamhafeez) [reply src/constant/constant.ts:21]: Okay

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (ghulamhafeez) [reply src/app/attendance/components/AttendanceFormBuilder.tsx:208]: Fix

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (ghulamhafeez) [reply src/app/attendance/components/AttendanceFormBuilder.tsx:248]: When you click on this input, showPicker() will open the date picker without requiring the user to click the calendar icon.

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (ghulamhafeez) [reply src/app/attendance/components/AttendanceFormBuilder.tsx:190]: Use the <Form /> component to build our form, as it’s a requirement for working with shadcn

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (ghulamhafeez) [reply src/app/attendance/components/AttendanceFormBuilder.tsx:181]: Fix

File: src/app/attendance/actions/attendanceActions.tsx
Comment (ghulamhafeez) [reply src/app/attendance/actions/attendanceActions.tsx:7]: Yes, fix

File: src/app/attendance/actions/attendanceActions.test.tsx
Comment (ghulamhafeez) [reply src/app/attendance/actions/attendanceActions.test.tsx:14]: fix

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (ghulamhafeez) [reply src/app/attendance/components/AttendanceFormBuilder.tsx:206]: No, the default value is a number; we submit the member ID, not the name. However, in the UI, we display the member name in the select

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (ghulamhafeez) [reply src/app/attendance/components/AttendanceFormBuilder.tsx:226]: it expecting string 
image

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (ghulamhafeez) [reply src/app/attendance/components/AttendanceFormBuilder.tsx:221]: Fix

File: src/app/attendance/actions/attendanceActions.test.tsx
Comment (ghulamhafeez) [reply src/app/attendance/actions/attendanceActions.test.tsx]: Fix

File: src/app/attendance/actions/CreateAttendance.tsx
Comment (Monfernape) [inline src/app/attendance/actions/CreateAttendance.tsx:1]: file name needs to be lowercase. Let's call it `create-attendance.ts`. Also, it's not tsx, its ts.

File: src/app/attendance/actions/CreateAttendance.tsx
Comment (Monfernape) [inline src/app/attendance/actions/CreateAttendance.tsx:8]: Typo again. It's Request not Requset

Comment (Monfernape) [inline src/app/attendance/actions/CreateAttendance.tsx:8]: Let's call it `createAttendance`

File: src/app/attendance/actions/createAttendance.test.tsx
Comment (Monfernape) [inline src/app/attendance/actions/createAttendance.test.tsx:37]: Nice

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (Monfernape) [reply src/app/attendance/components/AttendanceFormBuilder.tsx:189]: But why HTML form? Isn't ShadCN's `Form` enough?

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (Monfernape) [reply src/app/attendance/components/AttendanceFormBuilder.tsx:205]: We should do the mapping before looping.

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (Monfernape) [reply src/app/attendance/components/AttendanceFormBuilder.tsx:221]: `roleBaseMembers.map(x => ({...x, id: x.id.toString()}))`

**Review by Monfernape — APPROVED** [review]
Approving to unblock but let's fix few more things before merging this. Looks good though.

**Review by nadirhussain786 — APPROVED** [review]

File: src/app/attendance/actions/CreateAttendance.tsx
Comment (ghulamhafeez) [reply src/app/attendance/actions/CreateAttendance.tsx:8]: Fixed

File: src/app/attendance/actions/CreateAttendance.tsx
Comment (ghulamhafeez) [reply src/app/attendance/actions/CreateAttendance.tsx:1]: Fixed

File: src/app/attendance/components/AttendanceFormBuilder.tsx
Comment (ghulamhafeez) [reply src/app/attendance/components/AttendanceFormBuilder.tsx:221]: Fixed

**Pull request merged**
